packages/vite/src/node/plugins/resolve.ts

```diff
@@ -556,36 +579,12 @@ export function tryNodeResolve(
   const lastArrowIndex = id.lastIndexOf('>')
   const nestedRoot = id.substring(0, lastArrowIndex).trim()
   const nestedPath = id.substring(lastArrowIndex + 1).trim()
-
-  const possiblePkgIds: string[] = []
-  for (let prevSlashIndex = -1; ; ) {
-    let slashIndex = nestedPath.indexOf('/', prevSlashIndex + 1)
-    if (slashIndex < 0) {
-      slashIndex = nestedPath.length
-    }
-
-    const part = nestedPath.slice(
-      prevSlashIndex + 1,
-      (prevSlashIndex = slashIndex)
-    )
-    if (!part) {
-      break
-    }
-
-    // Assume path parts with an extension are not package roots, except for the
-    // first path part (since periods are sadly allowed in package names).
-    // At the same time, skip the first path part if it begins with "@"
-    // (since "@foo/bar" should be treated as the top-level path).
-    if (possiblePkgIds.length ? path.extname(part) : part[0] === '@') {
-      continue
-    }
-
-    const possiblePkgId = nestedPath.slice(0, slashIndex)
-    possiblePkgIds.push(possiblePkgId)
-  }
+  const nestedPkgId = nestedPath
+    .split('/', nestedPath[0] === '@' ? 2 : 1)
+    .join('/')
 
   let basedir: string
-  if (dedupe?.some((id) => possiblePkgIds.includes(id))) {
```

Comment:
No need for `possiblePkgIds` anymore with the new logic, as long we assume the `resolve.dedupe` array only ever references package IDs and not module IDs. This assumption makes sense to me!